masterslave.q

```diff
@@ -1,62 +1,88 @@
+ };
 
 \d .masterslave
 
 //table scehma of connected lb processes with ismaster status
-statustable:([handle:`int$()] procname:`symbol$();starttimeUTC:`timestamp$();ismaster:`boolean$())
+statustable:([handle:`int$()] procname:();starttimeUTC:`timestamp$();ismaster:`boolean$());
 
 //store own process start timestamp
-start:.z.p
+start:.z.p;
 
 //set details dict with own details
-details:{`procname`starttimeUTC`ismaster!(.proc.procname;.masterslave.start;1b)} 
+details:{
+ `procname`starttimeUTC!(string .proc.procname;.masterslave.start)
+ };
 
 //get details of procname provided
-getdetails:{[processname] (first exec w from .servers.SERVERS where procname like processname) ".masterslave.details[]"}
+getdetails:{[processname]
+ (first exec w from .servers.SERVERS where procname like processname,not null w) (`.masterslave.details;[])
+ };
 
-//update .masterslave.statustable with other proc details and update ismaster col to determine which process is master 
+//update .masterslave.statustable with other proc details and update ismaster col to determine which process is master
 addmember:{[processname]
-    update ismaster:0b from (`.masterslave.statustable upsert .masterslave.getdetails[processname],(enlist `handle)!(enlist first exec w from .servers.SERVERS  where procname like processname)) where starttimeUTC<>min starttimeUTC}
+ `.masterslave.statustable upsert .masterslave.getdetails[processname],
+  (`handle`ismaster)!(first exec w from .servers.SERVERS  where procname like processname,not null w;1b)
+ };
+
+masterupdate:{update ismaster:starttimeUTC=min starttimeUTC from `.masterslave.statustable}
 
 //is the process itself the master
-ammaster:{exec ismaster from .masterslave.statustable where handle=0}
+ammaster:{exec ismaster from .masterslave.statustable where handle=0};
 
-//find which process is the master
-findmaster:{first exec handle from .masterslave.statustable where ismaster=1b}
+//find which process is the master - can only be run after checkifmaster has been run
+findmaster:{exec handle, procname from .masterslave.statustable where ismaster=1b};
 
 //wrapper func for finding master
 checkifmaster:{
-    .servers.startup[];
-    .masterslave.addmember each string exec procname from .servers.SERVERS where proctype like "lb", not null w
- }
+ (.masterslave.addmember')[string exec procname from .servers.SERVERS where proctype like "lb", not null w];
+ .masterslave.masterupdate[]
+ };
+
+deletedropped:{[W]delete from `.masterslave.statustable where handle=W};
 
 //add pc override here
 //remove dropped connection from statustable
-//tell other alive processes to renegotiate who the master is 
-
+//tell other alive processes to renegotiate who the master is
+pc:{[result;W]
+ .masterslave.deletedropped[W];
+ .masterslave.checkifmaster[];
+ result
+ };
 
 \d .
 
 //populate statutable and find which process is master
-.masterslave.checkifmaster[]
-
+init[];
 
-//.z.pc call 
+//.z.pc call
+.z.pc:{.masterslave.pc[x y;y]}.z.pc;
```

Comment:
need to wrap this in an error trap (in case .z.pc isn't initially defined) i.e. 
.z.pc:{.masterslave.pc[x y;y]}@[value;`.z.pc;{{[x]}}]
Also as per comment above, better to have .masterslave.pc only take one param, and then .z.pc is 

.z.pc:{x y; .masterslave.pc[y]}@[value;`.z.pc;{{[x]}}]

.z.pc should only be defined within init[]